lucene/core/src/java/org/apache/lucene/search/ScoreMode.java

```diff
@@ -51,10 +59,51 @@ public boolean needsScores() {
     public boolean needsScores() {
       return true;
     }
+    @Override
+    public boolean isExhaustive() {
+      return false;
+    }
+  },
+
+  /**
+   * ScoreMode for top field collectors that can provide their own iterators,
+   * to optionally allow to skip for non-competitive docs
+   */
+  TOP_DOCS {
+    @Override
+    public boolean needsScores() {
+      return false;
+    }
+    @Override
+    public boolean isExhaustive() {
+      return false;
+    }
+  },
+
+  /**
+   * ScoreMode for top field collectors that can provide their own iterators,
+   * to optionally allow to skip for non-competitive docs.
+   * This mode is used when there is a secondary sort by _score.
+   */
+  TOP_DOCS_WITH_SCORES {
+    @Override
+    public boolean needsScores() {
+      return true;
+    }
+    @Override
+    public boolean isExhaustive() {
+      return false;
+    }
   };
 
   /**
    * Whether this {@link ScoreMode} needs to compute scores.
    */
   public abstract boolean needsScores();
+
+  /**
+   * Returns {@code true} if for this {@link ScoreMode} it is necessary to process all documents,
+   * or {@code false} if is enough to go through top documents only.
+   */
+  public abstract boolean isExhaustive();
```

Comment:
Rather than defining these as abstract, maybe have two final fields set in a constructor?  Then the definitions become much less verbose: `TOP_DOCS_WITH_SCORES(true, false);`

Reply:
@romseygeek Thanks, a very good suggestion, this indeed made the code more clear and less verbose.  Addressed in 719882e